File: db/csharp/SalesManagement.Tests/IntegrationTests/ProductTests.cs
using FluentAssertions;
using SalesManagement.Domain.Models;
using SalesManagement.Infrastructure.Repositories;
using Xunit;

namespace SalesManagement.Tests.IntegrationTests
{
    /// <summary>
    /// 商品マスタのテスト
    /// </summary>
    public class ProductTests : DatabaseTestBase
    {
        [Fact]
        public async Task 商品を登録できる()
        {
            // Arrange
            var categoryRepo = new ProductCategoryRepository(ConnectionString);
            var productRepo = new ProductRepository(ConnectionString);

            // 前提：商品分類が存在
            var category = CreateTestCategory("CAT001", "電子機器");
            await categoryRepo.InsertAsync(category);

            var product = CreateTestProduct("PROD001", "ノートパソコン A型", "CAT001");

            // Act
            await productRepo.InsertAsync(product);

            // Assert
            var found = await productRepo.FindByIdAsync("PROD001");
            found.Should().NotBeNull();
            found!.ProductFormalName.Should().Be("ノートパソコン A型");
            found.SellingPrice.Should().Be(150000);
        }

        [Fact]
        public async Task 商品分類コードで商品を検索できる()
        {
            // Arrange
            var categoryRepo = new ProductCategoryRepository(ConnectionString);
            var productRepo = new ProductRepository(ConnectionString);

            var category = CreateTestCategory("CAT001", "電子機器");
            await categoryRepo.InsertAsync(category);

            await productRepo.InsertAsync(CreateTestProduct("PROD001", "ノートPC A型", "CAT001"));
            await productRepo.InsertAsync(CreateTestProduct("PROD002", "ノートPC B型", "CAT001"));

            // Act
            var products = (await productRepo.FindByCategoryAsync("CAT001")).ToList();

            // Assert
            products.Should().HaveCount(2);
        }

        [Fact]
        public async Task 商品を更新できる()
        {
            // Arrange
            var categoryRepo = new ProductCategoryRepository(ConnectionString);
            var productRepo = new ProductRepository(ConnectionString);

            var category = CreateTestCategory("CAT001", "電子機器");
            await categoryRepo.InsertAsync(category);

            var product = CreateTestProduct("PROD001", "ノートPC A型", "CAT001");
            await productRepo.InsertAsync(product);

            // Act
            product.ProductFormalName = "ノートPC A型 改訂版";
            product.SellingPrice = 160000;
            product.UpdatedAt = DateTime.Now;
            product.UpdatedBy = "updater";
            await productRepo.UpdateAsync(product);

            // Assert
            var updated = await productRepo.FindByIdAsync("PROD001");
            updated.Should().NotBeNull();
            updated!.ProductFormalName.Should().Be("ノートPC A型 改訂版");
            updated.SellingPrice.Should().Be(160000);
        }

        [Fact]
        public async Task 商品を削除できる()
        {
            // Arrange
            var categoryRepo = new ProductCategoryRepository(ConnectionString);
            var productRepo = new ProductRepository(ConnectionString);

            var category = CreateTestCategory("CAT001", "電子機器");
            await categoryRepo.InsertAsync(category);

            var product = CreateTestProduct("PROD001", "ノートPC A型", "CAT001");
            await productRepo.InsertAsync(product);

            // Act
            await productRepo.DeleteAsync("PROD001");

            // Assert
            var deleted = await productRepo.FindByIdAsync("PROD001");
            deleted.Should().BeNull();
        }

        // テストデータ作成ヘルパーメソッド
        private static ProductCategory CreateTestCategory(string code, string name)
        {
            return new ProductCategory
            {
                ProductCategoryCode = code,
                ProductCategoryName = name,
                ProductCategoryLevel = 1,
                ProductCategoryPath = code,
                LowestLevelFlag = 1,
                CreatedAt = DateTime.Now,
                CreatedBy = "admin",
                UpdatedAt = DateTime.Now,
                UpdatedBy = "admin"
            };
        }

        private static Product CreateTestProduct(string code, string name, string categoryCode)
        {
            return new Product
            {
                ProductCode = code,
                ProductFormalName = name,
                ProductAbbreviation = name,
                ProductNameKana = "ノートパソコン",
                ProductType = "PRODUCT",
                ModelNumber = "MODEL-001",
                SellingPrice = 150000,
                PurchasePrice = 100000,
                CostOfSales = 100000,
                TaxType = 1,
                ProductCategoryCode = categoryCode,
                MiscellaneousType = 0,
                InventoryManagementFlag = 1,
                InventoryAllocationFlag = 0,
                SupplierCode = null,
                SupplierBranch = null,
                CreatedAt = DateTime.Now,
                CreatedBy = "admin",
                UpdatedAt = DateTime.Now,
                UpdatedBy = "admin"
            };
        }
    }
}

File: db/fsharp/SalesManagement.Tests/IntegrationTests/ProductTests.fs
module SalesManagement.Tests.IntegrationTests.ProductTests

open System
open Xunit
open FsUnit.Xunit
open SalesManagement.Domain.Models
open SalesManagement.Infrastructure.Repositories
open SalesManagement.Infrastructure.Repositories.ProductRepository
open SalesManagement.Tests.DatabaseTestBase

type ProductTests() =
    inherit DatabaseTestBase()

    let createTestCategory code name =
        {
            ProductCategoryCode = code
            ProductCategoryName = name
            ProductCategoryLevel = 1
            ProductCategoryPath = code
            LowestLevelFlag = 1
            CreatedAt = DateTime.Now
            CreatedBy = "admin"
            UpdatedAt = DateTime.Now
            UpdatedBy = "admin"
        }

    let createTestProduct code name categoryCode =
        {
            ProductCode = code
            ProductFormalName = name
            ProductAbbreviation = name
            ProductNameKana = "テストショウヒン"
            ProductType = "製品"
            ModelNumber = Some "MODEL-001"
            SellingPrice = 150000
            PurchasePrice = 100000
            CostOfSales = 100000
            TaxType = 1
            ProductCategoryCode = categoryCode
            MiscellaneousType = 0
            InventoryManagementFlag = 1
            InventoryAllocationFlag = 0
            SupplierCode = None
            SupplierBranch = None
            CreatedAt = DateTime.Now
            CreatedBy = "admin"
            UpdatedAt = DateTime.Now
            UpdatedBy = "admin"
        }

    [<Fact>]
    member this.``商品を登録できる``() =
        task {
            // Arrange
            let category = createTestCategory "CAT001" "電子機器"
            do! ProductCategoryRepository.insertAsync this.ConnectionString category

            let product = createTestProduct "PROD001" "ノートパソコン A型" "CAT001"

            // Act
            do! insertAsync this.ConnectionString product

            // Assert
            let! found = findByIdAsync this.ConnectionString "PROD001"
            found |> should not' (equal None)
            found.Value.ProductFormalName |> should equal "ノートパソコン A型"
            found.Value.SellingPrice |> should equal 150000
        }

    [<Fact>]
    member this.``商品分類コードで商品を検索できる``() =
        task {
            // Arrange
            let category = createTestCategory "CAT001" "電子機器"
            do! ProductCategoryRepository.insertAsync this.ConnectionString category

            do! insertAsync this.ConnectionString (createTestProduct "PROD001" "ノートPC A型" "CAT001")
            do! insertAsync this.ConnectionString (createTestProduct "PROD002" "ノートPC B型" "CAT001")

            // Act
            let! products = findByCategoryAsync this.ConnectionString "CAT001"
            let productList = products |> Seq.toList

            // Assert
            productList |> should haveLength 2
            productList |> List.forall (fun p -> p.ProductCategoryCode = "CAT001") |> should be True
        }

    [<Fact>]
    member this.``商品を更新できる``() =
        task {
            // Arrange
            let category = createTestCategory "CAT001" "電子機器"
            do! ProductCategoryRepository.insertAsync this.ConnectionString category

            let product = createTestProduct "PROD001" "ノートPC A型" "CAT001"
            do! insertAsync this.ConnectionString product

            // Act
            let updated =
                { product with
                    ProductFormalName = "ノートPC A型 改良版"
                    SellingPrice = 180000
                    UpdatedAt = DateTime.Now
                    UpdatedBy = "updater" }
            do! updateAsync this.ConnectionString updated

            // Assert
            let! result = findByIdAsync this.ConnectionString "PROD001"
            result |> should not' (equal None)
            result.Value.ProductFormalName |> should equal "ノートPC A型 改良版"
            result.Value.SellingPrice |> should equal 180000
        }
